File: app/src/main/java/com/jnjcomu/edison/broadcast/PlengiEventBroadcastReceiver.java
package com.jnjcomu.edison.broadcast;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.support.annotation.NonNull;
import android.widget.Toast;

import com.jnjcomu.edison.api.APIBuilder;
import com.jnjcomu.edison.storage.UserStorage;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class PlengiEventBroadcastReceiver extends BroadcastReceiver {
    public static final String RECEIVER_ID = "com.jnjcomu.edison.cloud.response";

    @Override
    public void onReceive(Context context, Intent intent) {
        if (!intent.getAction().equals(RECEIVER_ID)) return;
        Toast.makeText(context, intent.getStringExtra("place.id"), Toast.LENGTH_SHORT).show();

        noticeRegion(context, intent, true);
    }

    private void noticeRegion(Context context, Intent intent, boolean retry) {
        UserStorage userStorage = UserStorage.getInstance(context);

        APIBuilder.getAPI().enter(
                userStorage.getTicket(),
                intent.getStringExtra("place.id")
        ).enqueue(new Callback<Void>() {
            @Override
            public void onResponse(@NonNull Call<Void> call, @NonNull Response<Void> response) {
                if (!response.isSuccessful()) {
                    APIBuilder.getAPI()
                            .fetchTicket(userStorage.getTicket())
                            .subscribe(ticket -> {
                                UserStorage.getInstance(context).saveUserTicket(ticket);
                            });

                    if (retry) noticeRegion(context, intent, false);
                }
            }

            @Override
            public void onFailure(@NonNull Call<Void> call, @NonNull Throwable t) {
            }
        });
    }
}

File: app/src/main/java/com/jnjcomu/edison/broadcast/PlengiEventBroadcastReceiver.kt
package com.jnjcomu.edison.broadcast

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import com.jnjcomu.edison.addition.EdisonAPI
import com.jnjcomu.edison.addition.userStorage
import com.loplat.placeengine.PlengiResponse
import org.jetbrains.anko.toast
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

class PlengiEventBroadcastReceiver : BroadcastReceiver() {

    override fun onReceive(context: Context, intent: Intent) {
        if (intent.action != RECEIVER_ID) return

        val enterType = intent.getIntExtra("response.enterType", 0)
        if (enterType == PlengiResponse.EnterType.ENTER)
            context.toast("${intent.getStringExtra("place.name")}에 입실했습니다.")
        else if (enterType == PlengiResponse.EnterType.NEARBY)
            context.toast("${intent.getStringExtra("place.name")} 주변에 있습니다.")

        noticeRegion(context, intent, true)
    }

    private fun noticeRegion(context: Context, intent: Intent, retry: Boolean) {
        EdisonAPI.enter(
                context.userStorage.ticket!!,
                intent.getStringExtra("place.id")
        ).enqueue(object : Callback<Void> {
            override fun onResponse(call: Call<Void>, response: Response<Void>) {
                if (!response.isSuccessful) {
                    EdisonAPI
                            .fetchTicket(context.userStorage.ticket!!)
                            .subscribe { ticket -> context.userStorage.saveUserTicket(ticket) }

                    if (retry) noticeRegion(context, intent, false)
                }
            }

            override fun onFailure(call: Call<Void>, t: Throwable) {}

        })
    }

    companion object {
        val RECEIVER_ID = "com.jnjcomu.edison.cloud.response"
    }
}
